sweep/sweeper.go

```diff
@@ -1,40 +1,691 @@
+		// listener for spend and schedule a sweep.
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners = append(
+					pendInput.listeners, input.resultChan)
+				continue
+			}
+
+			// Create a new pendingInput and initialize the
+			// listeners slice with the passed in result channel. If
+			// this input is offered for sweep again, the result
+			// channel will be appended to this slice.
+			pendInput = &pendingInput{
+				listeners:        []chan Result{input.resultChan},
+				input:            input.input,
+				minPublishHeight: bestHeight,
+			}
+			s.pendingInputs[outpoint] = pendInput
+
+			// Start watching for spend of this input, either by us
+			// or the remote party.
+			cancel, err := s.waitForSpend(
+				outpoint,
+				input.input.SignDesc().Output.PkScript,
+				input.input.HeightHint(),
+			)
+			if err != nil {
+				err = fmt.Errorf(
+					"cannot watch for spend: %v", err,
+				)
+
+				s.signalAndRemove(&outpoint, Result{Err: err})
+
+				continue
+			}
+			pendInput.ntfnRegCancel = cancel
+
+			// Check to see if with this new input a sweep tx can be
+			// formed.
+			s.scheduleSweep(bestHeight)
+
+		// A spend of one of our inputs is detected. Signal sweep
+		// results to the caller(s).
+		case spend := <-s.spendChan:
+			// For testing
+			if s.testSpendChan != nil {
+				s.testSpendChan <- *spend.SpentOutPoint
+			}
+
+			// Query store to find out if we every published this
+			// tx.
+			spendHash := *spend.SpenderTxHash
+			isOurTx, err := s.cfg.Store.IsOurTx(spendHash)
+			if err != nil {
+				log.Errorf("Cannot determine if tx %v "+
+					"is ours: %v", spendHash, err)
+				continue
+			}
+
+			// Signal sweep results for inputs in this confirmed tx.
+			for _, txIn := range spend.SpendingTx.TxIn {
+				outpoint := txIn.PreviousOutPoint
+
+				// Check if this input is known to us. It could
+				// probably be unknown if we canceled the
+				// registration, deleted from pendingInputs but
+				// the ntfn was in-flight already. Or this could
+				// be not one of our inputs.
+				_, ok := s.pendingInputs[outpoint]
+				if !ok {
+					continue
+				}
+
+				// Return either a nil or a remote spend result.
+				var err error
+				if !isOurTx {
+					err = ErrRemoteSpend
+				}
+
+				// Signal result channels.
+				s.signalAndRemove(&outpoint, Result{
+					Tx:  spend.SpendingTx,
+					Err: err,
+				})
+			}
+
+			// Now that an input of ours is spent, we can try to
+			// resweep the remaining inputs.
+			s.scheduleSweep(bestHeight)
+
+		// The timer expires and we are going to (re)sweep.
+		case <-s.timer:
+			log.Debugf("Timer expired")
+
+			// Set timer to nil so we know that a new timer needs to
+			// be started when new inputs arrive.
+			s.timer = nil
```

Comment:
If we use the `ticker` package here, then we can utilize the `Pause()` and `Resume` methods.